### PR TITLE
fix(dynamic-links,android): getInitialLink returned more than once, and sometimes returned null

### LINKED ISSUE #3027: 🔥 Clear getInitialLink after first call
## Issue

Just making an issue for this as it was part of another issue and was getting lost. 

`getInitialLink` will return the link until the app is killed. There are certain scenarios where you need to process the link on the first time it shows up only, it would be great if it returned null after the first call (or cleared the link in some manner). @mikehardy [was talking about dequeing in this comment ](https://github.com/invertase/react-native-firebase/issues/2618#issuecomment-543884753) seems like it would work great here.

### Javascript

<details><summary>Click To Expand</summary>
<p>

### iOS

<details><summary>Click To Expand</summary>
<p>

#### `ios/Podfile`:

- [ ] I'm not using Pods
- [x] I'm using Pods and my Podfile looks like:

#### `AppDelegate.m`:

</p>
</details>

---

### Android

<details><summary>Click To Expand</summary>
<p>

#### Have you converted to AndroidX?

- [x] my application is an AndroidX application?
- [x] I am using `android/gradle.settings` `jetifier=true` for Android compatibility?
- [ ] I am using the NPM package `jetifier` for react-native compatibility?

#### `AndroidManifest.xml`:

---

## Environment

- **Platform that you're experiencing the issue on**:
  - [ ] iOS
  - [ ] Android
  - [ ] **iOS** but have not tested behavior on Android
  - [ ] **Android** but have not tested behavior on iOS
  - [x] Both
- **`react-native-firebase` version you're using that has this issue:**
  - `e.g. 5.6.0`
- **`Firebase` module(s) you're using that has the issue:**
  - `DynamicLinks`
- **Are you using `TypeScript`?**
  - `Y  3.7.3`
  
</p>
</details>

---

Think `react-native-firebase` is great? Please consider supporting all of the project maintainers and contributors by donating via our [Open Collective](https://opencollective.com/react-native-firebase/donate) where all contributors can submit expenses. [[Learn More]](https://invertase.io/oss/react-native-firebase/contributing/donations-expenses)

- 👉 Check out [`React Native Firebase`](https://twitter.com/rnfirebase) and [`Invertase`](https://twitter.com/invertaseio) on Twitter for updates on the library.

### LINKED ISSUE #3990: 🔥  Incorrect/Same dynamic link received in 'getInitialLink' except the first time when the app in the background state.
## Issue

I am receiving Incorrect/Same dynamic link received in 'getInitialLink' except the first time when the app in the background state.
---

### Javascript

<details><summary>Click To Expand</summary>
<p>

#### `package.json`:

```json
{
  "name": "ReactNativeProject",
  "version": "0.0.1",
  "private": true,
  "scripts": {
    "android": "react-native run-android",
    "ios": "react-native run-ios",
    "start": "react-native start",
    "test": "jest",
    "lint": "eslint ."
  },
  "dependencies": {
    "@react-native-community/async-storage": "^1.7.1",
    "@react-native-community/datetimepicker": "^2.1.2",
    "@react-native-community/geolocation": "^2.0.2",
    "@react-native-community/masked-view": "^0.1.5",
    "@react-native-community/netinfo": "^5.3.3",
    "@react-native-community/picker": "^1.1.3",
    "@react-native-firebase/analytics": "^6.7.1",
    "@react-native-firebase/app": "^6.7.1",
    "@react-native-firebase/crashlytics": "^6.7.1",
    "@react-native-firebase/dynamic-links": "^6.7.1",
    "@react-native-firebase/messaging": "^6.7.1",
    "appcenter": "2.6.1",
    "appcenter-analytics": "2.6.1",
    "appcenter-crashes": "2.6.1",
    "aws-amplify": "^2.2.7",
    "aws-amplify-react-native": "^3.2.2",
    "moment": "^2.24.0",
    "react": "16.9.0",
    "react-native": "0.61.5",
    "react-native-android-location-services-dialog-box": "^2.8.2",
    "react-native-banner-carousel": "^1.0.3",
    "react-native-camera": "^3.21.0",
    "react-native-code-push": "^6.0.0",
    "react-native-collapsible": "^1.5.1",
    "react-native-country-picker-modal": "^1.10.0",
    "react-native-gesture-handler": "^1.5.3",
    "react-native-image-crop-picker": "^0.31.1",
    "react-native-image-picker": "^2.3.1",
    "react-native-image-resizer": "^1.2.2",
    "react-native-keyboard-spacer": "^0.4.1",
    "react-native-map-link": "^2.7.4",
    "react-native-qrcode-scanner": "^1.4.0",
    "react-native-read-more-text": "^1.1.2",
    "react-native-reanimated": "^1.6.0",
    "react-native-responsive-screen": "^1.4.1",
    "react-native-safe-area-context": "^0.6.2",
    "react-native-screens": "^2.0.0-alpha.32",
    "react-native-simple-toast": "^1.0.0",
    "react-native-star-rating": "^1.1.0",
    "react-native-swiper-flatlist": "^2.0.3",
    "react-native-tab-view": "^2.11.0",
    "react-native-timeline-flatlist": "^0.6.5",
    "react-native-vector-icons": "6.6.0",
    "react-navigation": "^4.0.10",
    "react-navigation-drawer": "^2.3.3",
    "react-navigation-stack": "^2.0.1",
    "rn-fetch-blob": "^0.12.0",
    "underscore": "^1.9.2"
  },
  "devDependencies": {
    "@babel/core": "7.7.7",
    "@babel/runtime": "7.7.7",
    "@react-native-community/eslint-config": "0.0.5",
    "babel-jest": "24.9.0",
    "eslint": "6.8.0",
    "jest": "24.9.0",
    "metro-react-native-babel-preset": "0.56.3",
    "react-test-renderer": "16.9.0"
  },
  "jest": {
    "preset": "react-native"
  },
  "babel": {
    "ignore": [
      "node_modules/aws-sdk/dist/aws-sdk-react-native.js"
    ]
  }
}

```

#### `firebase.json` for react-native-firebase v6:

```json
# N/A
```

</details>

### iOS

<details><summary>Click To Expand</summary>
<p>

#### `ios/Podfile`:

- [ ] I'm not using Pods
- [x] I'm using Pods and my Podfile looks like:

```ruby
# N/A
```

#### `AppDelegate.m`:

```objc
// N/A
```

</p>
</details>

---

### Android

<details><summary>Click To Expand</summary>
<p>

#### Have you converted to AndroidX?

- [x] my application is an AndroidX application?
- [x] I am using `android/gradle.settings` `jetifier=true` for Android compatibility?
- [ ] I am using the NPM package `jetifier` for react-native compatibility?

#### `android/build.gradle`:

```groovy
// Top-level build file where you can add configuration options common to all sub-projects/modules.

buildscript {
    ext {
        buildToolsVersion = "28.0.3"
        minSdkVersion = 16
        compileSdkVersion = 28
        targetSdkVersion = 28
        androidXCore = "1.0.2"
    }
    repositories {
        google()
        jcenter()
        maven {
            url 'https://maven.fabric.io/public'
        }
    }
    dependencies {
        classpath("com.android.tools.build:gradle:3.4.2")
        classpath 'com.google.gms:google-services:4.3.3'
        classpath 'io.fabric.tools:gradle:1.31.2'
        // NOTE: Do not place your application dependencies here; they belong
        // in the individual module build.gradle files
    }
}

allprojects {
    repositories {
        mavenLocal()
        maven {
            // All of React Native (JS, Obj-C sources, Android binaries) is installed from npm
            url("$rootDir/../node_modules/react-native/android")
        }
        maven {
            // Android JSC is installed from npm
            url("$rootDir/../node_modules/jsc-android/dist")
        }

        google()
        jcenter()
        maven { url 'https://jitpack.io' }
    }
}

```

#### `android/app/build.gradle`:

```groovy
apply plugin: "com.android.application"
apply plugin: 'io.fabric'
import com.android.build.OutputFile

/**
 * The react.gradle file registers a task for each build variant (e.g. bundleDebugJsAndAssets
 * and bundleReleaseJsAndAssets).
 * These basically call `react-native bundle` with the correct arguments during the Android build
 * cycle. By default, bundleDebugJsAndAssets is skipped, as in debug/dev mode we prefer to load the
 * bundle directly from the development server. Below you can see all the possible configurations
 * and their defaults. If you decide to add a configuration block, make sure to add it before the
 * `apply from: "../../node_modules/react-native/react.gradle"` line.
 *
 * project.ext.react = [
 *   // the name of the generated asset file containing your JS bundle
 *   bundleAssetName: "index.android.bundle",
 *
 *   // the entry file for bundle generation
 *   entryFile: "index.android.js",
 *
 *   // https://facebook.github.io/react-native/docs/performance#enable-the-ram-format
 *   bundleCommand: "ram-bundle",
 *
 *   // whether to bundle JS and assets in debug mode
 *   bundleInDebug: false,
 *
 *   // whether to bundle JS and assets in release mode
 *   bundleInRelease: true,
 *
 *   // whether to bundle JS and assets in another build variant (if configured).
 *   // See http://tools.android.com/tech-docs/new-build-system/user-guide#TOC-Build-Variants
 *   // The configuration property can be in the following formats
 *   //         'bundleIn${productFlavor}${buildType}'
 *   //         'bundleIn${buildType}'
 *   // bundleInFreeDebug: true,
 *   // bundleInPaidRelease: true,
 *   // bundleInBeta: true,
 *
 *   // whether to disable dev mode in custom build variants (by default only disabled in release)
 *   // for example: to disable dev mode in the staging build type (if configured)
 *   devDisabledInStaging: true,
 *   // The configuration property can be in the following formats
 *   //         'devDisabledIn${productFlavor}${buildType}'
 *   //         'devDisabledIn${buildType}'
 *
 *   // the root of your project, i.e. where "package.json" lives
 *   root: "../../",
 *
 *   // where to put the JS bundle asset in debug mode
 *   jsBundleDirDebug: "$buildDir/intermediates/assets/debug",
 *
 *   // where to put the JS bundle asset in release mode
 *   jsBundleDirRelease: "$buildDir/intermediates/assets/release",
 *
 *   // where to put drawable resources / React Native assets, e.g. the ones you use via
 *   // require('./image.png')), in debug mode
 *   resourcesDirDebug: "$buildDir/intermediates/res/merged/debug",
 *
 *   // where to put drawable resources / React Native assets, e.g. the ones you use via
 *   // require('./image.png')), in release mode
 *   resourcesDirRelease: "$buildDir/intermediates/res/merged/release",
 *
 *   // by default the gradle tasks are skipped if none of the JS files or assets change; this means
 *   // that we don't look at files in android/ or ios/ to determine whether the tasks are up to
 *   // date; if you have any other folders that you want to ignore for performance reasons (gradle
 *   // indexes the entire tree), add them here. Alternatively, if you have JS files in android/
 *   // for example, you might want to remove it from here.
 *   inputExcludes: ["android/**", "ios/**"],
 *
 *   // override which node gets called and with what additional arguments
 *   nodeExecutableAndArgs: ["node"],
 *
 *   // supply additional arguments to the packager
 *   extraPackagerArgs: []
 * ]
 */

project.ext.react = [
    entryFile: "index.js",
    enableHermes: false,  // clean and rebuild if changing
]

apply from: "../../node_modules/react-native/react.gradle"
apply from: "../../node_modules/react-native-code-push/android/codepush.gradle"

/**
 * Set this to true to create two separate APKs instead of one:
 *   - An APK that only works on ARM devices
 *   - An APK that only works on x86 devices
 * The advantage is the size of the APK is reduced by about 4MB.
 * Upload all the APKs to the Play Store and people will download
 * the correct one based on the CPU architecture of their device.
 */
def enableSeparateBuildPerCPUArchitecture = false

/**
 * Run Proguard to shrink the Java bytecode in release builds.
 */
def enableProguardInReleaseBuilds = true

/**
 * The preferred build flavor of JavaScriptCore.
 *
 * For example, to use the international variant, you can use:
 * `def jscFlavor = 'org.webkit:android-jsc-intl:+'`
 *
 * The international variant includes ICU i18n library and necessary data
 * allowing to use e.g. `Date.toLocaleString` and `String.localeCompare` that
 * give correct results when using with locales other than en-US.  Note that
 * this variant is about 6MiB larger per architecture than default.
 */
def jscFlavor = 'org.webkit:android-jsc:+'

/**
 * Whether to enable the Hermes VM.
 *
 * This should be set on project.ext.react and mirrored here.  If it is not set
 * on project.ext.react, JavaScript will not be compiled to Hermes Bytecode
 * and the benefits of using Hermes will therefore be sharply reduced.
 */
def enableHermes = project.ext.react.get("enableHermes", false);

android {
    compileSdkVersion rootProject.ext.compileSdkVersion

    compileOptions {
        sourceCompatibility JavaVersion.VERSION_1_8
        targetCompatibility JavaVersion.VERSION_1_8
    }

    defaultConfig {
        applicationId "com.reactnative.project"
        minSdkVersion rootProject.ext.minSdkVersion
        targetSdkVersion rootProject.ext.targetSdkVersion
        versionCode 5
        versionName "1.1.0"
        missingDimensionStrategy 'react-native-camera', 'general'
    }
    splits {
        abi {
            reset()
            enable enableSeparateBuildPerCPUArchitecture
            universalApk false  // If true, also generate a universal APK
            include "armeabi-v7a", "x86", "arm64-v8a", "x86_64"
        }
    }
    signingConfigs {
        debug {
            storeFile file('debug.keystore')
            storePassword 'android'
            keyAlias 'androiddebugkey'
            keyPassword 'android'
        }
        release {
            storeFile file('debug.keystore')
            storePassword 'android'
            keyAlias 'androiddebugkey'
            keyPassword 'android'
        }
    }
    buildTypes {
        debug {
            signingConfig signingConfigs.debug
            multiDexEnabled true
        }
        release {
            // Caution! In production, you need to generate your own keystore file.
            // see https://facebook.github.io/react-native/docs/signed-apk-android.
            signingConfig signingConfigs.release
            minifyEnabled enableProguardInReleaseBuilds
            proguardFiles getDefaultProguardFile("proguard-android.txt"), "proguard-rules.pro"
        }
    }
    // applicationVariants are e.g. debug, release
    applicationVariants.all { variant ->
        variant.outputs.each { output ->
            // For each separate APK per architecture, set a unique version code as described here:
            // https://developer.android.com/studio/build/configure-apk-splits.html
            def versionCodes = ["armeabi-v7a": 1, "x86": 2, "arm64-v8a": 3, "x86_64": 4]
            def abi = output.getFilter(OutputFile.ABI)
            if (abi != null) {  // null for the universal-debug, universal-release variants
                output.versionCodeOverride =
                        versionCodes.get(abi) * 1048576 + defaultConfig.versionCode
            }

        }
    }
}

dependencies {
    implementation fileTree(dir: "libs", include: ["*.jar"])
    implementation "com.facebook.react:react-native:+"  // From node_modules

    if (enableHermes) {
        def hermesPath = "../../node_modules/hermes-engine/android/";
        debugImplementation files(hermesPath + "hermes-debug.aar")
        releaseImplementation files(hermesPath + "hermes-release.aar")
    } else {
        implementation jscFlavor
    }
}

// Run this once to be able to run the application with BUCK
// puts all compile dependencies into folder libs for BUCK to use
task copyDownloadableDepsToLibs(type: Copy) {
    from configurations.compile
    into 'libs'
}

apply from: file("../../node_modules/@react-native-community/cli-platform-android/native_modules.gradle"); applyNativeModulesAppBuildGradle(project)
apply from: "../../node_modules/react-native-vector-icons/fonts.gradle"
apply plugin: 'com.google.gms.google-services'

```

#### `android/settings.gradle`:

```groovy
rootProject.name = 'ReactNativeProject'
apply from: file("../node_modules/@react-native-community/cli-platform-android/native_modules.gradle"); applyNativeModulesSettingsGradle(settings)
include ':app'

```

#### `MainApplication.java`:

```java
// N/A
```

#### `AndroidManifest.xml`:

```xml
<manifest xmlns:android="http://schemas.android.com/apk/res/android"
  package="com.reactnative.project">

    <uses-permission android:name="android.permission.INTERNET" />
    <uses-permission android:name="android.permission.ACCESS_NETWORK_STATE" />
    <uses-permission android:name="android.permission.CAMERA" />
    <uses-permission android:name="android.permission.READ_EXTERNAL_STORAGE" />
    <uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE" />


    <application
      android:name=".MainApplication"
      android:label="@string/app_name"
      android:icon="@mipmap/ic_launcher"
      android:roundIcon="@mipmap/ic_launcher_round"
      android:allowBackup="false"
      android:theme="@style/AppTheme"
      android:networkSecurityConfig="@xml/network_security_config">
      <activity
        android:name=".SplashActivity"
        android:theme="@style/SplashTheme"
        android:label="@string/app_name"
        android:launchMode="singleTask">
        <intent-filter>
            <action android:name="android.intent.action.MAIN" />
            <category android:name="android.intent.category.LAUNCHER" />
        </intent-filter>
      </activity>
      <activity
        android:name=".MainActivity"
        android:label="@string/app_name"
        android:configChanges="keyboard|keyboardHidden|orientation|screenSize"
        android:windowSoftInputMode="adjustResize"
        android:exported="true"
        android:launchMode="singleTask"/>
      <activity android:name="com.facebook.react.devsupport.DevSettingsActivity" />
    </application>

</manifest>

```

</p>
</details>

---

## Environment

<details><summary>Click To Expand</summary>
<p>

**`react-native info` output:**

```
info Fetching system and libraries information...
System:
    OS: macOS Mojave 10.14.5
    CPU: (4) x64 Intel(R) Core(TM) i5-5350U CPU @ 1.80GHz
    Memory: 24.82 MB / 8.00 GB
    Shell: 3.2.57 - /bin/bash
  Binaries:
    Node: 10.16.3 - /usr/local/bin/node
    npm: 6.9.0 - /usr/local/bin/npm
    Watchman: 4.9.0 - /usr/local/bin/watchman
  SDKs:
    iOS SDK:
      Platforms: iOS 13.2, DriverKit 19.0, macOS 10.15, tvOS 13.2, watchOS 6.1
  IDEs:
    Android Studio: 3.5 AI-191.8026.42.35.6010548
    Xcode: 11.3/11C29 - /usr/bin/xcodebuild
  npmPackages:
    react: 16.9.0 => 16.9.0 
    react-native: 0.61.5 => 0.61.5 
  npmGlobalPackages:
    react-native-cli: 2.0.1
    react-native-create-library: 3.1.2
    react-native-rename: 2.4.1

```

- **Platform that you're experiencing the issue on**:
  - [ ] iOS
  - [ ] Android
  - [ ] **iOS** but have not tested behavior on Android
  - [x] **Android** but have not tested behavior on iOS
  - [ ] Both
- **`react-native-firebase` version you're using that has this issue:**
  - `e.g. 5.4.3`
- **`Firebase` module(s) you're using that has the issue:**
  - `e.g. Instance ID`
- **Are you using `TypeScript`?**
  - `Y/N` & `VERSION`

</p>
</details>

---

- 👉 Check out [`React Native Firebase`](https://twitter.com/rnfirebase) and [`Invertase`](https://twitter.com/invertaseio) on Twitter for updates on the library.

### PR DESCRIPTION
**Pull request by sharc7**
### Description

This PR addresses two issue:
1. `getInitialLink ` link is returned more than once.
2. `getInitialLink ` returns null.
* The Android SDK doesn't seem to clear the dynamic link after calling getDynamicLink in `getInitialLink`.  
(as stated in the official docs: `Calling getDynamicLink() retrieves the link and clears that data so it is only processed once by your app.`)  
This causes the link to be retrieved even when returning to the app from the Overview screen after backgrounding the app using the back button, which leads to unwanted results such as deep links being handled more than once.
* When `getInitialLink` is called before the host is resumed, `getCurrentActivity` will be null and `getInitialLink` will return null instead of trying to `getDynamicLink`.

### Related issues

Fixes #3990 
Fixes #3027 
### Checklist

- I read the [Contributor Guide](../CONTRIBUTING.md) and followed the process outlined there for submitting PRs.
  - [x] Yes
- My change supports the following platforms;
  - [x] `Android`
  - [ ] `iOS`
- My change includes tests;
  - [ ] `e2e` tests added or updated in `packages/\*\*/e2e`
  - [ ] `jest` tests added or updated in `packages/\*\*/__tests__`
- [ ] I have updated TypeScript types that are affected by my change.
- This is a breaking change;
  - [ ] Yes
  - [x] No

### Test Plan

---

Think `react-native-firebase` is great? Please consider supporting the project with any of the below:

- 👉 Star this repo on GitHub ⭐️
- 👉 Follow [`React Native Firebase`](https://twitter.com/rnfirebase) and [`Invertase`](https://twitter.com/invertaseio) on Twitter

:fire:

### CODE REVIEW

**Comment by mikehardy** [comment]
Hi there! Thanks for posting this, it looks like it is solving a real problem. However I have some clarifying questions in order to understand it since the code is not immediately obvious

> This PR addresses two issue:
> 
>    getInitialLink link is returned more than once.

Am I correct in understanding that the first issue was already addressed but the previous solution is just incomplete for a specific condition? That condition being:

>    getInitialLink returns null.

>    The Android SDK doesn't seem to clear the dynamic link after calling getDynamicLink in getInitialLink.
>    (as stated in the official docs: Calling getDynamicLink() retrieves the link and clears that data so it is only processed once by your app.)

The language used here makes this seem like it is not handled at all but in order to understand the PR I need to be sure that I'm understanding: we already attempt to handle this firebase-ios-sdk behavior in a way react-native-firebase-consumers want (get data then clear it), just that it is not working 100%. Otherwise I'm not the `gotInitialLink` variable that existed before in the code

>    This causes the link to be retrieved even when returning to the app from the Overview screen after backgrounding the app using the back button, which leads to unwanted results such as deep links being handled more than once.
>    When getInitialLink is called before the host is resumed, getCurrentActivity will be null and getInitialLink will return null instead of trying to getDynamicLink.

Assuming I understand correctly that the module already attempts to handle it, this part is not clear to me.

So if I understand this correctly: 

- if app not in background and initial link not fetched, "works" (link marked as fetched, link returned)
- if app in background and initial link already fetched, "works" (link already marked as fetched, no link returned)
- if app in background and initial link not fetched, fails - this PR fixes this? (link not marked as fetched, no link returned)

What if link is marked as fetched, app is in background, and user opens a distinct / different link? Should the initialLinkUrl itself be stored and used as the canary (instead of a gotInitialLink boolean) and then compared to see if it's different? 🤔

**Comment by sharc7** [comment]
Hi, thanks for the quick reply!
> Am I correct in understanding that the first issue was already addressed but the previous solution is just incomplete for a specific condition? That condition being on app background / resume:

Yes that's correct, mostly for the special case that is "backgrounded using the back button".

> The language used here makes this seem like it is not handled at all but in order to understand the PR I need to be sure that > I'm understanding: we already attempt to handle this firebase-ios-sdk behavior in a way react-native-firebase consumers want (get data then clear it), just that it is not working 100%. Otherwise I'm not sure why the gotInitialLink variable that existed before in the code

> Assuming I understand correctly that the module already attempts to handle it, this part is not clear to me.
> 
> So if I understand this correctly:
> 
> if app not in background and initial link not fetched, "works" (link marked as fetched, link returned)
> if app in background and initial link already fetched, "works" (link already marked as fetched, no link returned)
> if app in background and initial link not fetched, fails - this PR fixes this? (link not marked as fetched, no link returned)

I'm sorry, I didn't mean to say that it's not handled at all,  
just that in my experience the previous solution to the `getInitialLink link is returned more than once` issue was incomplete/not 100% working as it would return the initial url again when returning from the overview screen after the app is **backgrounded using the back button** or manually calling getInitialLink for example, in a touchable. that is - performing more than one call to `getInitialLink`.
Other backgrounding methods (home button or overview -> different app -> home button) worked as expected, meaning no link was returned because the host wasn't actually destroyed so `getInitialLink` wasn't called again when returning.

I've only managed to reproduce `getInitialLink returns null` issue after the onNewIntent was called and a link was returned, this sometimes took a few rounds of:
1. backgrounding using the home button
2. opening a link which resulted in the onNewIntent method to return a link (onLink)
3. backgrounding the app with the home button again
4. opening the same/different link
5. using the back button to background the app
6. finally opening the same or a different link

> What if link is marked as fetched, app is in background, and user opens a distinct / different link? Should the initialLinkUrl itself be stored and used as the canary (instead of a gotInitialLink boolean) and then compared to see if it's different? 🤔

I don't think so, if the app was _not_ backgrounded using the back button and then a link is opened, this will be handled by the onNewIntent hook (onLink).  
I don't think we should compare the actual link, just that know that a call was handled successfully.
In my opinion, and from what I understand from the docs, `getInitialLink` should only return a link once if it exists or null if doesn't, and should return null for all future calls (because the value should be cleared after the first call), unless it's the case where the app has been backgrounded by the back button _and_ a link launched the app, doesn't matter if it's a new or the same link, in this case (where the app is launched after a link was opened), we should try to return the link.
Other than that, I think that if the app (react-native, JS side) needs to use this value again after receiving it from react-native-firebase, it (the app) should manage this on its own.

I actually haven't thought about calling `getInitialLink` with a touchable before submitting the PR, so I've made some minor changes and addressed this issue with the `gotInitialLink` variable, I can push another commit if what I described above sounds good, or adjust according to what we discuss here before pushing.

**Comment by mikehardy** [comment]
Okay - thanks for the thoughtful reply

Can I ask you to clarify what you mean by "host"? In terms of react-native on Android I'm aware of the Android Application as a thing, I'm aware of the bridge, and I'm aware of the JS bundle running on the JS thread. When you say host, do you mean the Catalyst bridge? I think you do but I want to be clear - I don't think the actual words we use are that important I just have to make sure we're talking about the same concepts

Assuming this is correct, can you lay out (like you did in steps 1-6 above) the same or similar steps but this time also detailing for the error case exactly what state each of the 3 react-native app components is in (e.g. "Android Application Resumed, bridge up, original JS thread running", or "Android Application Paused, bridge down, JS thread destroyed", or "Android Application Resumed, bridge up, JS thread 2 running after bridge re-init" or similar)

I think you're on to something real here but I haven't probed this corner case and am not set up to reproduce it so I'll trust you on handling the corner cases etc

One thing I would like is a comment next to each member variable in the code explaining the state it is tracking, and then on the conditionals that test the new state variables, a comment explaining the state transition that we are looking for (and guarding against, basically)

**Comment by sharc7** [comment]
> Can I ask you to clarify what you mean by "host"? In terms of react-native on Android I'm aware of the Android Application as a thing, I'm aware of the bridge, and I'm aware of the JS bundle running on the JS thread. When you say host, do you mean the Catalyst bridge? I think you do but I want to be clear - I don't think the actual words we use are that important I just have to make sure we're talking about the same concepts

I'm not very familiar with React Native's (or Java's) internals, the reason I mentioned "host" was because the lifecycle method referred to "host" as in `onHostResume`, so whichever one that refers to.

>Assuming this is correct, can you lay out (like you did in steps 1-6 above) the same or similar steps but this time also detailing for the error case exactly what state each of the 3 react-native app components is in (e.g. "Android Application Resumed, bridge up, original JS thread running", or "Android Application Paused, bridge down, JS thread destroyed", or "Android Application Resumed, bridge up, JS thread 2 running after bridge re-init" or similar)
>
> I think you're on to something real here but I haven't probed this corner case and am not set up to reproduce it so I'll trust you on handling the corner cases etc

I'm not sure how to check all of those mentioned above, if you could assist in pointing me in the right direction I'd be happy to provide information.
In the meantime, I logged `getReactApplicationContext().getLifecycleState()` in `getInitialLink` and it appears that  
it returns `BEFORE_CREATE` or `BEFORE_RESUME` which is probably why `getCurrentActivity()` returns null at that point resulting at the method resolving with null.

> One thing I would like is a comment next to each member variable in the code explaining the state it is tracking, and then on the conditionals that test the new state variables, a comment explaining the state transition that we are looking for (and guarding against, basically)

Pushed a commit with some adjustments and comments as requested.

**Comment by mikehardy** [comment]
Cool - thanks for that - I'll re-queue this for review, that application state logging is key to understanding which part of the application (and what state it is in...) which should help me read through this and make sure I understand it. I mean, I think I do already but most the PRs here don't actually add state and state tracking to the system, being a glorified wrapper over an SDK makes for easy PRs normally, and I just want to make sure if I/we approve adding state that I/we get it right :-). Cheers

**Comment by sharc7** [comment]
cool, thanks!
I understand that adding state is unusual, there are probably other ways to solve this, so obviously feel free to change/discard this PR if it isn't correct, and let me know if there's anything else needed from me in this PR.
On a side note - thanks for all the hard work put into this project!

**Comment by mikehardy** [comment]
I think this is likely a case where adding state is the only way to handle it as the JS thread + bridge in react-native can have divergent state from native so it actually makes sense, I'm looking at it from the perspective of 99% sure I'm going to merge, possibly as-is from current PR state FWIW

My pleasure on the maintenance, I do enjoy making things work

**Comment by mikehardy** [comment]
Oh - the CLA must be signed though - please click the details link on that CI failure and do the needful for it to go green

**Comment by sharc7** [comment]
Sounds good. Thanks for the detailed explanation.

> Oh - the CLA must be signed though - please click the details link on that CI failure and do the needful for it to go green

Done 👍

File: packages/dynamic-links/android/src/main/java/io/invertase/firebase/dynamiclinks/ReactNativeFirebaseDynamicLinksModule.java [hunk 1]
```diff
@@ -40,9 +40,28 @@
...
-  private boolean gotInitialLink = false;
   private int initialLinkMinimumVersion = 0;
 
+  /**
+   * Ensures calls to getInitialLink only tries to retrieve the link from getDynamicLink once.
+   */
+  private boolean gotInitialLink = false;
+  /**
+   *  Used by getInitialLink to check if the activity has been resumed.
+   * "host" here refers to the host activity.
```
Comment (mikehardy) [inline packages/dynamic-links/android/src/main/java/io/invertase/firebase/dynamiclinks/ReactNativeFirebaseDynamicLinksModule.java:51]:
```suggestion
   * Used by getInitialLink to check if the activity has been resumed.
   * "host" refers to the host activity, in terms of {@link LifeCycleEventListener#onHostResume()}
```

File: packages/dynamic-links/android/src/main/java/io/invertase/firebase/dynamiclinks/ReactNativeFirebaseDynamicLinksModule.java [hunk 2]
```diff
@@ -106,25 +125,38 @@ public void buildShortLink(ReadableMap dynamicLinkMap, String shortLinkType, Pro
...
+    // Shouldn't happen anymore, left as a guard.
     if (currentActivity == null) {
       promise.resolve(null);
       return;
     }
 
-    FirebaseDynamicLinks.getInstance().getDynamicLink(currentActivity.getIntent())
+    Intent currentIntent = currentActivity.getIntent();
+    // Verify if the app was resumed from the Overview (history) screen.
+    launchedFromHistory = currentIntent != null && (currentIntent.getFlags() & Intent.FLAG_ACTIVITY_LAUNCHED_FROM_HISTORY) != 0;
```
Comment (mikehardy) [inline packages/dynamic-links/android/src/main/java/io/invertase/firebase/dynamiclinks/ReactNativeFirebaseDynamicLinksModule.java:153]:
```suggestion
    launchedFromHistory = (currentIntent != null) && ((currentIntent.getFlags() & Intent.FLAG_ACTIVITY_LAUNCHED_FROM_HISTORY) != 0);
```

This is pedantic, but I don't like forcing developers to think through precedence of operators, so I prefer explicit parentheses at all times

File: packages/dynamic-links/android/src/main/java/io/invertase/firebase/dynamiclinks/ReactNativeFirebaseDynamicLinksModule.java [hunk 3]
```diff
@@ -397,9 +434,21 @@ public void onActivityResult(Activity activity, int requestCode, int resultCode,
 
   @Override
   public void onHostResume() {
+    // Flag state is resumed.
+    hostResumed = true;
+    // Check if getInitialLink was called before LifeCycleState was RESUMED
+    // and there's a pending Promise.
+    if(initialPromise != null) {
```
Comment (mikehardy) [inline packages/dynamic-links/android/src/main/java/io/invertase/firebase/dynamiclinks/ReactNativeFirebaseDynamicLinksModule.java:441]:
```suggestion
    if (initialPromise != null) {
```

**Review by mikehardy — APPROVED** [review]
Nice!
I had only non-semantic changes (trivial whitespace etc) - this LGTM on a re-re-read ;-)

There was an unrelated timeout / CI failure on android which is unfortunate because this affects Android code. I'm waiting for a green CI run there then I can merge this

File: packages/dynamic-links/android/src/main/java/io/invertase/firebase/dynamiclinks/ReactNativeFirebaseDynamicLinksModule.java [hunk 4]
```diff
@@ -106,25 +125,38 @@ public void buildShortLink(ReadableMap dynamicLinkMap, String shortLinkType, Pro
...
-      } else {
-        promise.resolve(null);
-      }
+      promise.resolve(null);
+      return;
+    }
 
+    // Check for the case where getInitialLink
+    // runs before the LifeCycleState is RESUMED (e.g. BEFORE_CREATE or BEFORE_RESUME).
+    if(!hostResumed) {
```
Comment (mikehardy) [inline packages/dynamic-links/android/src/main/java/io/invertase/firebase/dynamiclinks/ReactNativeFirebaseDynamicLinksModule.java:137]:
```suggestion
    if (!hostResumed) {
```

**Comment by mikehardy** [comment]
CI went green! This should go out as 10.5.2 shortly, thanks!

**Comment by ivanstnsk** [comment]
Do you have the same behavior on iOS as well?

**Comment by mikehardy** [comment]
@ivanstnsk  The majority of this fix was peculiar to the interaction between android lifecycle / react-native design, with a sort of side fix of not returning the dynamic link multiple times. On iOS it is possible it still returns the same dynamic link multiple times if you call getInitialLink multiple times but I'm not 100%, have you tested it?

**Comment by ivanstnsk** [comment]
@mikehardy  I've discovered exactly the same behavior on iOS - the getInitialLink is not clearing the link after handling. Tested on @react-native-firebase/dynamic-links v10.5.1

**Comment by mikehardy** [comment]
@ivanstnsk I am not surprised - I'd welcome a PR to fix it and happily work with anyone that proposes them

**Comment by sharc7** [comment]
> CI went green! This should go out as 10.5.2 shortly, thanks!

Awesome, thank you! 😄 🚀
